mdbook/src/15-interrupts/nvic-and-interrupt-priority.md

```diff
@@ -0,0 +1,75 @@
+
+## Under The Hood
+
+We've seen that interrupts make our processor immediately jump to another function in the code, but
+what's going on behind the scenes to allow this to happen? In this section we'll cover some
+technical details that won't be necessary for the rest of the book, so feel free to skip ahead if
+you're not interested.
+
+### The Interrupt Controller
+
+Interrupts allow the processor to respond to peripheral events such as a GPIO input pin changing
+state, a timer completing its cycle, or a UART receiving a new byte. The peripheral contains
+circuitry that notices the event and informs a dedicated interrupt-handling peripheral. On Arm
+processors, the interrupt-handling peripheral is called the NVIC — the Nested Vector Interrupt
+Controller.
+
+> **NOTE** On other microcontroller architectures such as RISC-V the names and details discussed
+> here will differ, but the underlying principles are generally very similar.
+
+The NVIC can receive requests to trigger an interrupt from many peripherals. It's even common for a
+peripheral to have multiple possible interrupts, for example a GPIO port having an interrupt for
+each pin, or a UART having both a "data received" and "data finished transmission" interrupt. The
+job of the NVIC is to prioritise these interrupts, remember which ones still need to be procesed,
+and then cause the processor to run the relevant interrupt handler code.
+
+Depending on its configuration, the NVIC can ensure the current interrupt is fully processed before
+a new one is executed, or it can stop the processor in the middle of one interrupt in order to
+handle another that's higher priority.  This is called "preemption" and allows processors to respond
+very quickly to critical events.  For example, a robot controller might use low-priority interrupts
+to keep track sending status information to the operator, but also have a high-priority interrupt to
+detect an emergency stop button being pushed so it can immediately stop moving the motors. You
```

Comment:
Nitpicking here, but aren't emergency stop buttons (at least the nice big red ones) most of the time hooked straight into power supply lines? Sensor detects imminent collision and needs to interrupt whatever the program is doing might be more real in the sense intended here.